### PR TITLE
fix(dpmodel): align NativeLayer bias/idt init with pt MLPLayer

### PR DESCRIPTION
**Pull request by wanghan-iapcm**
## Summary

Two related changes addressing pt_expt's slow convergence vs pt baseline (~15-20% worse RMSE on multi-task DPA3) and the test-infrastructure brittleness it exposed.

### `ff6a9314` — fix(dpmodel): align NativeLayer bias/idt init with pt MLPLayer

`deepmd/dpmodel/utils/network.py` `NativeLayer.__init__` previously used Glorot scaling (`1/sqrt(num_in+num_out)`) for **all** of `w`, `b`, `idt`. `deepmd/pt`'s `MLPLayer._default_normal_init` uses Glorot only for `w`, with `b ~ N(0, 1)` and `idt ~ N(0.1, 0.001)`.

Adam preserves per-parameter scale, so the init mismatch persists through training and produces ~15-20% worse validation RMSE on multi-task DPA3 (e.g. `e_alloy` 1.51e-2 → 1.79e-2; `f_elec` 4.58e-2 → 5.53e-2 on a 500k-step run).

Fix aligns dpmodel with pt's well-tuned defaults; affects every dpmodel-based backend (pt_expt, jax, paddle) for newly-initialised models. Existing serialized models are unaffected (weights are overwritten on `deserialize`).

### `35d5cb21` — test: replace hardcoded C++ expected values with sidecar reference files

The init change above broke ~10 C++ test files that hardcoded `expected_e/f/v = {…}` initializer-list arrays manually pasted from `gen_*.py` script stdout. This refactor removes the manual-paste workflow:

- New header-only `source/api_cc/tests/expected_ref.h` parses plain-text sidecar files.
- New `gen_common.write_expected_ref()` writer; legacy `print_cpp_values` / `print_cpp_spin_values` retained as ad-hoc debug helpers.
- 7 gen scripts (dpa1/2/3, spin, fparam_aparam, model_devi) now emit `<model>.expected` sidecars at run time (each contains `[case]` sections with named arrays).
- 15 C++ tests load arrays via `ExpectedRef::get<VALUETYPE>(case, key)` in `SetUp()`. **No test cases removed** — every modified file preserves its original `TYPED_TEST` / `class TestInfer*` count vs upstream master.
- `.expected` files are gitignored (regenerated each CI run alongside the `.pt2`/`.pth`).
- `test_deeppot_ptexpt.cc` is intentionally left with its TF-derived hardcoded refs — its `.pt2` weights flow through `gen_fparam_aparam.py`'s `state_dict()` load path (not random init), so they're stable across this PR and any future numerics-perturbing change.
- `test_compressed_forward[float32]` for DPA1 / SeAttenV2: bump atol from `1e-5` to `4e-5` with a comment explaining the cause — the new bias init scale widens the embedding-net output range, observed compression-tabulation error ~3.2e-5.

## Test plan

- [x] `source/tests/common/dpmodel/test_network.py` — 20 passed
- [x] `source/tests/pt_expt/descriptor/` — 492 passed, 32 skipped (incl. the previously-failing float32 compression tests now passing at 4e-5)
- [x] `source/tests/common/` (skipping pre-existing TF/protobuf-broken `test_gui.py`) — 275 passed, 1 skipped
- [x] `source/tests/consistent/descriptor/` (with C++ tabulate ops built) — 2210 passed, 3678 skipped
- [x] `source/tests/consistent/{fitting,loss,model}/` + `test_type_embedding.py` — 3654 passed, 4736 skipped
- [x] C++ refactored tests (DPA1/2/3 pt+ptexpt, dpa_pt_spin, dpa_ptexpt_spin, fparam variants, default_fparam variants, model_devi Precomputed, spin_model_devi PtExpt) — **106 passed, 0 failed** locally with `gen_*.py` regenerated `.pt2/.pth/.expected`
- [ ] Empirical: re-run the 500k-step multi-task DPA3 training under pt_expt with this fix and confirm convergence matches pt baseline (out of scope for this PR — verified diagnostically only)

## Known limitation

The init alignment also affects jax / paddle backends through dpmodel. Their cross-backend consistency tests use `serialize`/`deserialize` round-trips and are unaffected, but newly-initialised jax / paddle models will pick up the new init too. If a downstream user relies on the old Glorot-bias behaviour, this is a breaking numerical change for new-from-scratch training (existing checkpoints unaffected).

## Summary by CodeRabbit

* **Refactor**
  * Adjusted neural network parameter initializations that may subtly change initial numeric/residual behavior.

* **Tests**
  * Replaced embedded large expected arrays with external sidecar reference files and centralized model-path constants across many tests.
  * Simplified test setup patterns and raised one float32 comparison tolerance to avoid spurious failures.

* **Chores**
  * Added support for standardized expected-value sidecar read/write and a .expected ignore rule.

### CODE REVIEW

File: deepmd/dpmodel/utils/network.py [hunk 1]
```diff
@@ -117,15 +117,16 @@ def __init__(
...
-        self.w = rng.normal(size=(num_in, num_out), scale=scale_factor).astype(prec)
-        self.b = (
-            rng.normal(size=(num_out,), scale=scale_factor).astype(prec)
-            if bias
-            else None
-        )
+        # Match deepmd/pt MLPLayer._default_normal_init: w uses Glorot scaling,
+        # but b is N(0, 1) (unscaled) and idt is N(0.1, 0.001). Using Glorot
+        # scaling for b/idt empirically slows convergence on ResNet-style
+        # fitting nets because residual layers start near identity (idt~0).
```
[the remark this answers is not in the document]
Comment (wanghan-iapcm) [reply deepmd/dpmodel/utils/network.py:123]: Not valid. The `idt~0` refers to the **Glorot** scaling we rejected (`N(0, 1/sqrt(num_in+num_out))`, mean=0), not the new init. The comment structure is: line 1 describes what the code does (w Glorot, b N(0,1), idt N(0.1, 0.001) matching pt); lines 2-4 explain *why* Glorot for b/idt is bad. Your suggestion would invert the meaning by attributing 'starts near identity' to the new (good) `idt~0.1` init.

File: source/api_cc/tests/test_deeppot_model_devi_ptexpt.cc [hunk 2]
```diff
@@ -451,6 +451,10 @@ class TestInferDeepPotModeDeviPtExptPrecomputed : public ::testing::Test {
 #if !defined(BUILD_PYTORCH) || !BUILD_PT_EXPT
     GTEST_SKIP() << "Skip because PyTorch support is not enabled.";
 #endif
+    deepmd_test::ExpectedRef ref;
+    ref.load("../../tests/infer/model_devi.expected");
+    expected_md_f = ref.get<VALUETYPE>("default", "expected_md_f");
+    expected_md_v = ref.get<VALUETYPE>("default", "expected_md_v");
     natoms = coord.size() / 3;
```
[the remark this answers is not in the document]
Comment (wanghan-iapcm) [reply source/api_cc/tests/test_deeppot_model_devi_ptexpt.cc:461]: Fixed in 00c45a6f — added `ASSERT_EQ(expected_md_f.size(), 3u)` and same for `expected_md_v` before the fixed-index access. The Precomputed test now fails cleanly if a future gen_model_devi.py change perturbs the sidecar layout.

File: source/tests/infer/gen_model_devi.py [hunk 3]
```diff
@@ -139,52 +140,10 @@ def main():
...
-        for ii, vv in enumerate(virial_flat):
-            comma = "," if ii < len(virial_flat) - 1 else ""
-            print(f"      {vv:.18e}{comma}")  # noqa: T201
-        print("  };")  # noqa: T201
-
     # ---- 4. Compute deviation stats (LAMMPS nlist) ----
+    # The C++ Precomputed test only consumes the model-deviation aggregate
+    # statistics (expected_md_f, expected_md_v); per-model atom_energy/force/
+    # virial are not asserted on the C++ side, so we don't write them.
     dp0 = DeepPot(models[0])
```
[the remark this answers is not in the document]
Comment (wanghan-iapcm) [reply source/tests/infer/gen_model_devi.py:147]: Not valid — `RUF059` (`unused-unpacked-variable`) is in the project's `tool.ruff.lint.ignore` list in pyproject.toml, so ruff does not flag this and CI doesn't fail. The unpacking style here matches other gen_*.py scripts; leaving as-is for consistency.

File: source/lmp/tests/expected_ref.py [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+        if line.startswith("[") and line.endswith("]"):
+            current_section = line[1:-1]
+            sections[current_section] = {}
+            continue
+        if current_section is None:
+            raise ValueError(f"array '{line}' before any [section] in {path}")
+        key, count = line.split()
+        n = int(count)
+        values = [float(lines[i + j].strip()) for j in range(n)]
+        i += n
```
[the remark this answers is not in the document]
Comment (wanghan-iapcm) [reply source/lmp/tests/expected_ref.py]: Fixed in 00c45a6f — explicit token-count check on the header line and bounds check on the payload, both with the source path in the message. Verified locally with synthetic malformed inputs.

File: source/lmp/tests/test_lammps_faparam_pt2.py [hunk 5]
```diff
@@ -24,43 +27,27 @@
...
+# Guarded with try/except because gen_fparam_aparam.py only runs when PyTorch
+# is built; matrices that disable PyTorch (e.g. paddle-only) skip the test in
+# setup_module but still load this file at pytest collection time.
+try:
+    _ref = read_expected_ref(ref_file)["default"]
+    expected_ae = _ref["expected_e"]
+    expected_e = np.sum(expected_ae)
+    expected_f = _ref["expected_f"].reshape(6, 3)
+except FileNotFoundError:
+    expected_ae = expected_e = expected_f = None
```
[the remark this answers is not in the document]
Comment (wanghan-iapcm) [reply source/lmp/tests/test_lammps_faparam_pt2.py]: Fixed in a4442510 — same inlining as in test_lammps_dpa3_pt2.py.

File: source/lmp/tests/test_lammps_dpa3_pt2.py [hunk 6]
```diff
@@ -21,103 +24,26 @@
...
+# setup_module but still load this file at pytest collection time.
+try:
+    _ref = read_expected_ref(ref_file)["pbc"]
+    expected_ae = _ref["expected_e"]
+    expected_e = np.sum(expected_ae)
+    expected_f = _ref["expected_f"].reshape(6, 3)
+    # LAMMPS uses opposite sign convention for virial vs DeepPot atom_virial.
+    expected_v = -_ref["expected_v"].reshape(6, 9)
+except FileNotFoundError:
+    expected_ae = expected_e = expected_f = expected_v = None
```
[the remark this answers is not in the document]
Comment (wanghan-iapcm) [reply source/lmp/tests/test_lammps_dpa3_pt2.py]: Fixed in a4442510 — inlined to `expected_e = float(np.sum(_ref["expected_e"]))` so the per-atom-energy array is no longer pinned to a top-level name; the FileNotFoundError fallback dropped `expected_ae` accordingly.

File: source/lmp/tests/test_lammps_dpa3_pt2_nopbc.py [hunk 7]
```diff
@@ -21,103 +24,26 @@
...
+# setup_module but still load this file at pytest collection time.
+try:
+    _ref = read_expected_ref(ref_file)["nopbc"]
+    expected_ae = _ref["expected_e"]
+    expected_e = np.sum(expected_ae)
+    expected_f = _ref["expected_f"].reshape(6, 3)
+    # LAMMPS uses opposite sign convention for virial vs DeepPot atom_virial.
+    expected_v = -_ref["expected_v"].reshape(6, 9)
+except FileNotFoundError:
+    expected_ae = expected_e = expected_f = expected_v = None
```
[the remark this answers is not in the document]
Comment (wanghan-iapcm) [reply source/lmp/tests/test_lammps_dpa3_pt2_nopbc.py]: Fixed in a4442510 — same inlining as in test_lammps_dpa3_pt2.py.

File: source/lmp/tests/expected_ref.py [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+                f"invalid array header '{line}' in {path}; expected '<name> <count>'"
+            )
+        key, count = parts
+        n = int(count)
+        if i + n > len(lines):
+            raise ValueError(
+                f"array '{key}' expects {n} values, but only {len(lines) - i} remain in {path}"
+            )
+        values = [float(lines[i + j].strip()) for j in range(n)]
+        i += n
```
[the remark this answers is not in the document]
Comment (wanghan-iapcm) [reply source/lmp/tests/expected_ref.py:44]: Fixed in d3704b9b — added explicit `n < 0` check that raises ValueError naming the array and source file before the bounds check, so a negative count terminates with a clear message instead of rewinding the read position.

**Review by njzjz — APPROVED** [review]

**Review by iProzd — APPROVED** [review]
